rfcs/react-components/convergence/event-handlers-event-type.md

```diff
@@ -0,0 +1,198 @@
+
+```ts
+import * as React from 'react';
+
+// ======= For an existing component =======
+{
+  type OpenPopoverEvents = KeyboardEvent | MouseEvent; // existing event type
+
+  type MyComponentElement = HTMLElement;
+
+  type OnOpenChangeData = (
+    | EventData<'click', React.MouseEvent<MyComponentElement>>
+    | EventData<'keydown', React.KeyboardEvent<MyComponentElement>>
+    | EventData<'scroll', Event>
+  ) & {
+    open: boolean;
+  };
+
+  type PopoverProps = {
+    /**
+     * @deprecated Use onOpenChange2 instead.
+     */
+    onOpenChange?: (e: OpenPopoverEvents, data: OnOpenChangeData) => void;
+
+    onOpenChange2?: EventHandler<OnOpenChangeData>;
+  };
+}
+
+// ======= For a new component =======
+{
+  type MyComponentElement = HTMLElement;
+
+  type OnSomeEventData = EventData<'click', React.MouseEvent<MyComponentElement>> & {
+    open: boolean;
+  };
+  // If one day we need to add more events, we can just add them to the union:
+  // type OnSomeEventData =
+  //   | EventData<'click', React.MouseEvent<MyComponentElement>>
+  //   | (EventData<'focus', React.FocusEvent<MyComponentElement>> & {
+  //       open: boolean;
+  //     });
+
+  type SomeProps = {
+    onSomeEvent?: EventHandler<OnSomeEventData>;
+  };
+}
+```
+
+- Pros 👍:
+
+  1. Ensures correct typing and it can accommodate future, unforeseen event types.
+  2. Currently we use unions to type events in cases where a callback may be triggered by multiple events. Consumers will need to use type assertion or type predict to use a strongly-typed event. But this solution provides an easier way:
+
+     ```ts
+     {
+       // Current way:
+       function isMouseEvent(e: MouseEvent | TouchEvent): e is MouseEvent {
+         return e instanceof MouseEvent; // This type check is just for demo purpose. There are issues with using `instanceof` in multi-window scenarios: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/instanceof#instanceof_and_multiple_realms
+       }
+       const onOpenChange = (e: MouseEvent | TouchEvent) => {
+         console.log(e.clientX); // ❌ type error - Property 'clientX' does not exist on type 'TouchEvent'
+         if (isMouseEvent(e)) {
+           console.log(e.clientX); // ✅
+         }
+       };
+     }
+
+     {
+       // Proposed way:
+       type OnSomeEventData =
+         | EventData<'click', React.MouseEvent<MyComponentElement>>
+         | (EventData<'focus', React.FocusEvent<MyComponentElement>> & {
+             open: boolean;
+           });
+       const onOpenChange = (_e: Event, data: OnSomeEventData) => {
+         console.log(data.event?.currentTarget); // base event data can be used without checking data.type
+         console.log(data.event?.clientX); // Typescript will give an error here as expected
```

Comment:
If we remove `| undefined` from the type, then can remove the `?`. Also (nit) include ✅ and ❌ for consistency with the comment below:
```suggestion
         console.log(data.event.currentTarget); // ✅ base event data can be used without checking data.type
         console.log(data.event.clientX); // ❌ Typescript will give an error here as expected
```